cpp/src/c_api/hits.cpp

```diff
@@ -113,40 +114,46 @@ struct hits_functor : public cugraph::c_api::abstract_functor {
       weight_t hub_score_differences{0};
       size_t number_of_iterations{0};
 
-#if 0
-      // FIXME:  Implementation will look something like this.
-      
-      if (initial_hubs_guess_ != nullptr) {
-        //
-        // Need to renumber initial_hubs_guess_vertices
-        // Need to shuffle and populate hubs
-        //
-        //  This is the original pagerank code, it will be sort of like this
-        renumber_ext_vertices<vertex_t, multi_gpu>(handle_,
-                                                   personalization_vertices_->as_type<vertex_t>(),
-                                                   personalization_vertices_->size_,
-                                                   number_map->data(),
-                                                   graph_view.get_local_vertex_first(),
-                                                   graph_view.get_local_vertex_last(),
-                                                   do_expensive_check_);
+      if (initial_hubs_guess_vertices_ != nullptr) {
+        rmm::device_uvector<vertex_t> guess_vertices(initial_hubs_guess_vertices_->size_,
+                                                     handle_.get_stream());
+        rmm::device_uvector<weight_t> guess_values(initial_hubs_guess_values_->size_,
+                                                   handle_.get_stream());
+
+        raft::copy(guess_vertices.data(),
+                   initial_hubs_guess_vertices_->as_type<vertex_t>(),
+                   guess_vertices.size(),
+                   handle_.get_stream());
+        raft::copy(guess_values.data(),
+                   initial_hubs_guess_values_->as_type<weight_t>(),
+                   guess_values.size(),
+                   handle_.get_stream());
+
+        cugraph::renumber_ext_vertices<vertex_t, multi_gpu>(handle_,
+                                                            guess_vertices.data(),
+                                                            guess_vertices.size(),
+                                                            number_map->data(),
+                                                            graph_view.get_local_vertex_first(),
+                                                            graph_view.get_local_vertex_last(),
+                                                            do_expensive_check_);
+
+        cugraph::detail::collect_vertex_values_to_local<vertex_t, weight_t, multi_gpu>(
+          handle_, guess_vertices, guess_values, hubs, graph_view.get_local_vertex_first());
```

Comment:
If I am not mistaken, this code will not work as intended and looks fragile.

First, `hubs` is not initialized, so if user provides `guess_vertices` only for a subset of vertices, hubs can hold garbage values. If `guess_vertices` has duplicates, I am concerned potential concurrent writes to the same location can happen, and not sure this is not an undefined behavior. But I don't see these requirements in the documentation of `collect_vertex_values_to_local`.

Second, if I am not mistaken, `collect_vertex_values_to_local` may not work as intended with renumbered vertices.

Third, we may update this function to take `guess_vertices` & `guess_values` as R-value input parameters and return `hubs`; I think this is more functional and easier to understand.

And we may be able to re-implement this in a bit different way.

Say, first shuffle guess_vertices & guess_values kv pairs to target GPUs; renumber (and here, we can renumber just using local `number_map`); then 0 initialize hubs then scatter. I think this will be more efficient.

Reply:
Yes, I think this would be more efficient.  I will refactor this as you suggested.

I will add a test that specifies initial values so that we can verify that it works on the conditions you mention.

Reply:
I don't see these updates yet (if I am not mistaken), so is this PR still targeting 22.04?

Reply:
Just pushed these changes about an hour ago.